[Meng An] iP

EvaderFati opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/EvaderFati/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :star_struck:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Duke {
   public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
   }
}
```

Commit:
Add Gradle support

EvaderFati committed:
Level-1 implementation

EvaderFati committed:
Level-2 implementation

EvaderFati committed:
Level-3 basic functionalities

EvaderFati committed:
Level-3: change static to non static & improve printing

EvaderFati committed:
Level-3: add comment and resolve indentation issues

EvaderFati committed:
Level-3: optimize code structure

EvaderFati committed:
Level-3: handle IndexOutOfBounds exception

EvaderFati committed:
Level-4 implementation

EvaderFati committed:
Level-4 add comments

EvaderFati committed:
Level-5 handle errors

EvaderFati committed:
Level-5 implementation

EvaderFati committed:
Level-6 implementation

EvaderFati committed:
add TextUiTesting

EvaderFati committed:
Add comments

EvaderFati committed:
Level-7: Save the tasks in hard disk

EvaderFati committed:
Support date with format uuuu-MM-dd

EvaderFati committed:
Merge pull request #1 from EvaderFati/branch-Level-8

Support date with format uuuu-MM-dd

EvaderFati committed:
Merge branch 'master' into branch-Level-7

EvaderFati committed:
Merge pull request #2 from EvaderFati/branch-Level-7

Level-7: Save the tasks in hard disk

EvaderFati committed:
Resolve issues after merging

cheekean5848 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,301 @@
+            }
+        } else if (splitted.length == 2 && splitted[0].equals("unmark")) {
+            try {
+                int num = Integer.parseInt(splitted[1]);
+                this.unmark(num-1);
+            } catch (NumberFormatException e) {
+                this.printMsg("\t Opps! Pls check your numbers.\n");
+            } catch (IndexOutOfBoundsException e) {
+                this.printMsg("\t Opps! The item you wanna unmark is out of bounds.\n");
+            }
+        } else if (splitted[0].equals("todo")) {
+            try {
+                this.addTask(TaskType.TODO, input);
+            } catch (DukeException e) {
+                this.printMsg(e.getMessage());
+            }
+        } else if (splitted[0].equals("event")) {
+            try {
+                this.addTask(TaskType.EVENT, input);
+            } catch (DukeException e) {
+                this.printMsg(e.getMessage());
+            }
+        } else if (splitted[0].equals("deadline")) {
+            try {
+                this.addTask(TaskType.DEADLINE, input);
+            } catch (DukeException e) {
+                this.printMsg(e.getMessage());
+            }
+        } else if (splitted.length == 2 && splitted[0].equals("delete")) {
+            try {
+                int num = Integer.parseInt(splitted[1]);
+                this.delete(num-1);
+            } catch (NumberFormatException e) {
+                this.printMsg("\t Opps! Pls check your numbers.\n");
+            } catch (IndexOutOfBoundsException e) {
+                this.printMsg("\t Opps! The item you wanna mark is out of bounds.\n");
+            }
+        } else {
+            this.printMsg("\t ☹ OOPS!!! I'm sorry, but I don't know what that means :-(\n");
+        }
```
Is it better to have your create an Exception class, and use a one try and catch block the catch the Exception?

cheekean5848 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,301 @@
+            }
+            t = new Todo(description);
+        } else if (type == TaskType.EVENT) {
+            timeIndex = input.indexOf(" /at ");
+            if (timeIndex == -1) { 
+                // cannot find '/at'
+                throw new DukeException("\t ☹ OOPS!!! The time of an event cannot be empty.\n");
+            }
+            try {
+                description = input.substring(6, timeIndex);
+                time = input.substring(timeIndex + 5);
+            } catch (StringIndexOutOfBoundsException e) {
+                throw new DukeException("\t ☹ OOPS!!! The description or time of an event cannot be empty.\n");
+            }
+            if (description.equals("") || time.equals("")) {
+                throw new DukeException("\t ☹ OOPS!!! The description or time of an event cannot be empty.\n");
+            }
+            t = new Event(description, this.strToDate(time));
+        } else { 
+            // deadline
+            timeIndex = input.indexOf(" /by ");
+            if (timeIndex == -1) { 
+                // cannot find '/at'
+                throw new DukeException("\t ☹ OOPS!!! The time of a deadline cannot be empty.\n");
+            }
+            try {
+                description = input.substring(9, timeIndex);
+                time = input.substring(timeIndex + 5);
+            } catch (StringIndexOutOfBoundsException e) {
+                throw new DukeException("\t ☹ OOPS!!! The description or time of a deadline cannot be empty.\n");
+            }
+            if (description.equals("") || time.equals("")) {
+                throw new DukeException("\t ☹ OOPS!!! The description or time of a deadline cannot be empty.\n");
+            }
+            t = new Deadline(description, this.strToDate(time));
+        }
+        this.lst.add(t);
+        this.printMsg("\t Got it. I've added this task:\n\t   " + t.toString()
+                + "\n\t Now you have " + this.lst.size() + " tasks in the list.\n");
+    }
```
No deep nesting, I like it

cheekean5848 reviewed (commented):
Overall is good, I like how you structure your code, maybe can try to implement a more OOP way.
Keep it going!

soloplxya commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,301 @@
+     * @throws FileNotFoundException TODO
+     * @throws IOException TODO
+     */
+    public void loadData() throws IOException {
```
Perhaps can consider shortening this method as it is a bit [long](https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods).

soloplxya commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Unmarks a task.
+     */
+	public void unmarkAsDone() {
```
Should the indentation here match the rest of the methods?

soloplxya commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,301 @@
+            f.createNewFile();
+        }
+
+        Scanner scanner = new Scanner(f);
```
I like how the main path is prominent in this method! 👍

soloplxya commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,301 @@
+            return;
+        } else if (input.equals("list")) {
+            this.list();
+        } else if (splitted.length == 2 && splitted[0].equals("mark")) {
```
Perhaps can consider making splitted[0] a variable since it is used quite often in this code section.

soloplxya reviewed (commented):
Overall, the code seems good to me other than some minor nit fixes! Good job! 👍

glennljw commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,49 @@
+        String status = isDone ? "1" : "0";
+        return " | " + status + " | " + this.description + "\n";
+    }
+}
```
Maybe Task could be defined as an abstract class?

glennljw commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,301 @@
+            try {
+                int num = Integer.parseInt(splitted[1]);
+                this.unmark(num-1);
+            } catch (NumberFormatException e) {
+                this.printMsg("\t Opps! Pls check your numbers.\n");
+            } catch (IndexOutOfBoundsException e) {
+                this.printMsg("\t Opps! The item you wanna unmark is out of bounds.\n");
+            }
+        } else if (splitted[0].equals("todo")) {
+            try {
+                this.addTask(TaskType.TODO, input);
+            } catch (DukeException e) {
+                this.printMsg(e.getMessage());
+            }
+        } else if (splitted[0].equals("event")) {
+            try {
+                this.addTask(TaskType.EVENT, input);
+            } catch (DukeException e) {
+                this.printMsg(e.getMessage());
+            }
+        } else if (splitted[0].equals("deadline")) {
+            try {
+                this.addTask(TaskType.DEADLINE, input);
+            } catch (DukeException e) {
+                this.printMsg(e.getMessage());
+            }
+        } else if (splitted.length == 2 && splitted[0].equals("delete")) {
+            try {
+                int num = Integer.parseInt(splitted[1]);
+                this.delete(num-1);
+            } catch (NumberFormatException e) {
+                this.printMsg("\t Opps! Pls check your numbers.\n");
+            } catch (IndexOutOfBoundsException e) {
+                this.printMsg("\t Opps! The item you wanna mark is out of bounds.\n");
+            }
+        } else {
+            this.printMsg("\t ☹ OOPS!!! I'm sorry, but I don't know what that means :-(\n");
+        }
+        this.run();
+    }
```
Can consider making both splitted.length and splitted[0] a local variable in the method.

glennljw reviewed (commented):
Overall great abstraction of code! Keep it up! :smile:

EvaderFati committed:
Implement OOP

EvaderFati committed:
Implement find

EvaderFati committed:
Merge pull request #3 from EvaderFati/branch-Level-9

Implement find

EvaderFati committed:
Change Parser class to non static

EvaderFati committed:
Add JUnit Tests

EvaderFati committed:
Add JUnit for Task.java

EvaderFati committed:
Add JavaDoc

EvaderFati committed:
Follow the Coding Standard

EvaderFati committed:
Merge pull request #4 from EvaderFati/branch-A-JavaDoc

Add JavaDoc

EvaderFati committed:
Merge pull request #5 from EvaderFati/branch-A-CodingStandard

Follow the Coding Standard

EvaderFati committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle
Add provided gradle wrapper files.

EvaderFati committed:
Merge pull request #6 from EvaderFati/branch-A-Gradle

Add Gradle support

EvaderFati committed:
Add Gradle support for JavaFX

EvaderFati committed:
Use FXML

EvaderFati committed:
Use controllers

EvaderFati committed:
Remove support for text UI.

Since GUI is implemented by the previous commit, the text UI is no
longer needed.

Altering the command usage infomation to better fit the GUI rather than
Text UI.

Let's print out the welcome message when the GUI starts.

Extracting the run() method in Duke.java because we are using the JavaFX
controller MainWindow.java to acquire user inputs and give feedbacks.

EvaderFati committed:
Use Assertions

The data file containing the task list might not be in the correct
format.

Adding Java assertions to ensure that the data is corrected loaded.

EvaderFati committed:
Improve code quality

EvaderFati committed:
Merge pull request #8 from EvaderFati/branch-A-Assertions

Use Assertions

EvaderFati committed:
Merge branch 'master' into branch-A-CodeQuality

EvaderFati committed:
Merge pull request #9 from EvaderFati/branch-A-CodeQuality

Improve code quality

EvaderFati committed:
Replace tabs and rewrite comments

EvaderFati committed:
Implement undo command

EvaderFati committed:
Merge pull request #10 from EvaderFati/branch-C-Undo

Implement undo command

EvaderFati committed:
Add user guide

EvaderFati committed:
Set theme jekyll-theme-midnight

EvaderFati committed:
Add Ui.png

EvaderFati committed:
Merge branch 'master' of github.com:EvaderFati/ip-1
